cmake/CPM.cmake

```diff
@@ -505,3 +509,62 @@ function(cpm_is_git_tag_commit_hash GIT_TAG RESULT)
     endif()
   endif()
 endfunction()
+
+macro(prettify_cpm_add_package OUT_VAR IS_IN_COMMENT)
+    set(oneValueArgs
+      NAME
+      FORCE
+      VERSION
+      GIT_TAG
+      DOWNLOAD_ONLY
+      GITHUB_REPOSITORY
+      GITLAB_REPOSITORY
+      GIT_REPOSITORY
+      SOURCE_DIR
+      DOWNLOAD_COMMAND
+      FIND_PACKAGE_ARGUMENTS
+      NO_CACHE
+      GIT_SHALLOW
+    )
+      set(multiValueArgs
+        OPTIONS
+      )
+
+      cmake_parse_arguments(CPM_AGS "" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+      foreach(oneArgName ${oneValueArgs})
+          if(DEFINED CPM_AGS_${oneArgName})
+              if(${IS_IN_COMMENT})
+                  string(APPEND ${OUT_VAR} "#")
+              endif()
+              if(${oneArgName} STREQUAL "SOURCE_DIR")
+                  string(REPLACE ${CMAKE_SOURCE_DIR} "\${CMAKE_SOURCE_DIR}" CPM_ARGS_${oneArgName} ${CPM_ARGS_${oneArgName}} )
+              endif()
```

Comment:
Probably we should use `CMAKE_CURRENT_SOURCE_DIR` here, for a more portable approach.

Reply:
`CMAKE_CURRENT_SOURCE_DIR` will depend on the location where the package was called, `CMAKE_SOURCE_DIR` is the path where is the root `CMakeLists.txt`. Since `SOURCE_DIR` is the location of the package i only resolve the "root path".

Reply:
Hmmmm that is a good point. The source dir will always change depending on if the current package is used as the root or a subdirectory. I can see how this can cause some unexpected behaviour. Maybe we simply shouldn't add any packages using `SOURCE_DIR` to the package lock?

Reply:
It is an attempt to make the `SOURCE_DIR` more dynamic, since if its not resolve a little bit, the variable will hold the full path of your system, and it will be unusable by other dev. I was hoping to try 1 or 2 term to resolve some part (like if its in `_dep` or under the root of the project) since with that it will be valid for everybody. For us we usually put most package near the root of the project to ease the developing, but I know usually it's under `<build folder>/_dep/...` .

Reply:
Yeah, I understand the use case, though this can go wrong when trying to use a package as a dependency that itself has relative imports using `SOURCE_DIR`. In most cases using `SOURCE_DIR` will be for first-party code, so imo we should be fine simply omitting it in that case.

Reply:
Not sure to understand completely, can you show me the suggestion.
 From what you say we should remove the resolution of the path for the `SOURCE_DIR`, or remove `SOURCE_DIR`?

Reply:
Never mind, I just saw that we are already ignoring any args, if the `SOURCE_DIR` parameter is specified. Here is the logic that adds packages to the lock:

```cmake
  if(CPM_PACKAGE_LOCK_ENABLED)
    if((CPM_ARGS_VERSION AND NOT CPM_ARGS_SOURCE_DIR) OR CPM_INCLUDE_ALL_IN_PACKAGE_LOCK)
      cpm_add_to_package_lock(${CPM_ARGS_NAME} "${ARGN}")
    elseif(CPM_ARGS_SOURCE_DIR)
      cpm_add_comment_to_package_lock(${CPM_ARGS_NAME} "local directory")
    else()
      cpm_add_comment_to_package_lock(${CPM_ARGS_NAME} "${ARGN}")
    endif()
  endif()
```

Which makes sense, as adding `SOURCE_DIR` to the package lock can introduce unexpected bugs. This also means that the prettify function can happily ignore the `SOURCE_DIR` argument.

Reply:
ok I will add add the ignore step for `SOURCE_DIR`

Reply:
To be clear I think we can just delete lines 540-542 as Packages using SOURCE_DIR won't have the argument here anyways.

Reply:
mmm, could we could simply add a test like:
```
cpm_prettify_package_arguments(PRETTY_ARGN true "local directory")
set(EXPECTED_COMMENTED_LOCALDIR
    "#    local directory
"
)
assert_equal(${PRETTY_ARGN} ${EXPECTED_COMMENTED_LOCALDIR})
```
since if we define: `CPM_INCLUDE_ALL_IN_PACKAGE_LOCK` the `SOURCE_DIR` will be added to the lock file.

Reply:
this test will cover the `elseif(CPM_ARGS_SOURCE_DIR)` part.

Reply:
> since if we define: `CPM_INCLUDE_ALL_IN_PACKAGE_LOCK` the `SOURCE_DIR` will be added to the lock file.

Ok gotcha, I wasn't aware that you're explicitly enabling it. As it's disabled by default (it shouldn't be used for dependencies anyways) I guess the original check against `CMAKE_SOURCE_DIR` is fine after all.